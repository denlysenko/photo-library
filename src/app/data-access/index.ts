export * from './photos/photos.service';
