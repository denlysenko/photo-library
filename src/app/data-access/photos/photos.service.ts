import { Injectable } from '@angular/core';

import faker from 'faker';

import { Observable, of } from 'rxjs';
import { delay } from 'rxjs/operators';

import { FAVORITES_KEY } from '../../constants';
import { StorageService } from '../../core';
import { Photo } from '../../interfaces';

function generateRandomInt(min: number, max: number): number {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

function generatePhotos(count = 12): Photo[] {
  const photos = [];

  for (let i = 0; i < count; i++) {
    photos.push({
      id: faker.random.uuid(),
      link: faker.image.image(),
    });
  }

  return photos;
}

type PhotoEntities = {
  [id: string]: Photo;
};

@Injectable({ providedIn: 'root' })
export class PhotosService {
  constructor(private readonly storageService: StorageService) {}

  getPhotos(): Observable<Photo[]> {
    return of(generatePhotos()).pipe(delay(generateRandomInt(200, 300)));
  }

  getAllFavorites(): Observable<Photo[]> {
    const favorites = this.storageService.getItem<PhotoEntities>(FAVORITES_KEY);

    return of(favorites === null ? [] : Object.values(favorites)).pipe(
      delay(generateRandomInt(200, 300))
    );
  }

  getFavoriteById(id: string): Observable<Photo> {
    const favorites = this.storageService.getItem<PhotoEntities>(FAVORITES_KEY);

    return of(favorites === null ? null : favorites[id]).pipe(
      delay(generateRandomInt(200, 300))
    );
  }

  addToFavorites(photo: Photo): Observable<boolean> {
    const favorites = this.storageService.getItem<PhotoEntities>(FAVORITES_KEY);

    this.storageService.setItem<PhotoEntities>(FAVORITES_KEY, {
      ...favorites,
      [photo.id]: photo,
    });

    return of(true).pipe(delay(generateRandomInt(200, 300)));
  }

  removeFromFavorites(id: string): Observable<boolean> {
    const favorites = this.storageService.getItem<PhotoEntities>(FAVORITES_KEY);

    if (favorites !== null) {
      const { [id]: removed, ...updatedPhotos } = favorites;
      this.storageService.setItem(FAVORITES_KEY, updatedPhotos);
    }

    return of(true).pipe(delay(generateRandomInt(200, 300)));
  }
}
