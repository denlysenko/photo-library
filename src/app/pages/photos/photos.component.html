<app-infinite-scroll (scrolled)="onScrolled()">
  <app-photo-grid [photos]="photos$ | async"
                  (photoClicked)="onPhotoClick($event)"></app-photo-grid>
  <app-inline-spinner [class.hidden]="(loading$ | async) === false"></app-inline-spinner>
</app-infinite-scroll>