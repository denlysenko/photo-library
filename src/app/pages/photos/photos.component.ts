import { ChangeDetectionStrategy, Component } from '@angular/core';
import { ActivatedRoute } from '@angular/router';

import { merge, Observable, Subject } from 'rxjs';
import {
  catchError,
  distinctUntilChanged,
  map,
  switchMap,
  tap,
} from 'rxjs/operators';

import { ProgressComponent } from '../../base';
import { DEFAULT_LIMIT } from '../../constants';
import { PhotosService } from '../../data-access';
import { Photo } from '../../interfaces';

@Component({
  selector: 'app-photos',
  templateUrl: './photos.component.html',
  styleUrls: ['./photos.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class PhotosComponent extends ProgressComponent {
  readonly loadMore = new Subject<void>();

  private cache: Photo[] = [];

  readonly photos$: Observable<Photo[]> = merge(
    this.route.data.pipe(map((data) => data.photos)),
    this.loadMore.asObservable().pipe(
      switchMap(() => {
        this.loading.next(true);
        return this.photosService.getPhotos().pipe(
          tap(() => this.loading.next(false)),
          catchError(() => {
            this.loading.next(false);
            return [];
          })
        );
      })
    )
  ).pipe(
    distinctUntilChanged(),
    map((photos) => {
      this.cache = [...this.cache, ...photos];
      return this.cache;
    })
  );

  constructor(
    private readonly route: ActivatedRoute,
    private readonly photosService: PhotosService
  ) {
    super();
  }

  onScrolled(): void {
    if (this.cache.length <= DEFAULT_LIMIT) {
      return;
    }

    this.loadMore.next();
  }

  onPhotoClick(photo: Photo): void {
    this.photosService.addToFavorites(photo).subscribe();
  }
}
