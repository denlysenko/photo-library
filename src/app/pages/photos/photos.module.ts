import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';

import {
  InfiniteScrollModule,
  InlineSpinnerModule,
  PhotoGridModule,
} from '../../ui';
import { PhotosComponent } from './photos.component';
import { PhotosResolver } from './photos.resolver';

@NgModule({
  imports: [
    CommonModule,
    PhotoGridModule,
    InfiniteScrollModule,
    InlineSpinnerModule,
  ],
  declarations: [PhotosComponent],
  providers: [PhotosResolver],
})
export class PhotosModule {}
