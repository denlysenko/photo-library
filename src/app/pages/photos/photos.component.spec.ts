import { CommonModule } from '@angular/common';
import {
  async,
  ComponentFixture,
  fakeAsync,
  TestBed,
  tick,
} from '@angular/core/testing';
import { ActivatedRoute } from '@angular/router';

import { of } from 'rxjs';

import { PhotosService } from '../../data-access';
import {
  InfiniteScrollModule,
  InlineSpinnerModule,
  PhotoGridModule,
} from '../../ui';
import { PhotosComponent } from './photos.component';

(window as any).IntersectionObserver = class IntersectionObserver {
  constructor() {}

  observe(): void {
    return null;
  }

  disconnect(): void {
    return null;
  }
};

const photos = [
  {
    id: 'uuid',
    link: 'link',
  },
];

describe('PhotosComponent', () => {
  let component: PhotosComponent;
  let fixture: ComponentFixture<PhotosComponent>;
  let photosService: PhotosService;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [
        CommonModule,
        PhotoGridModule,
        InfiniteScrollModule,
        InlineSpinnerModule,
      ],
      declarations: [PhotosComponent],
      providers: [
        {
          provide: ActivatedRoute,
          useValue: {
            data: of({
              photos,
            }),
          },
        },
        {
          provide: PhotosService,
          useValue: {
            getPhotos: jest.fn(() =>
              of([
                {
                  id: 'uuid2',
                  link: 'link2',
                },
              ])
            ),
            addToFavorites: jest.fn(() => of(true)),
          },
        },
      ],
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(PhotosComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();

    photosService = TestBed.inject(PhotosService);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should create', () => {
    expect(fixture).toMatchSnapshot();
  });

  describe('loadMore', () => {
    it('should call getPhotos() and render new photos', fakeAsync(() => {
      component.loadMore.next();
      tick();
      fixture.detectChanges();
      expect(fixture).toMatchSnapshot();
      expect(photosService.getPhotos).toHaveBeenCalledTimes(1);
    }));
  });

  describe('onPhotoClick', () => {
    it('should call addToFavorites()', fakeAsync(() => {
      fixture.nativeElement.querySelector('[data-testid=photo-tile]').click();
      tick();
      expect(photosService.addToFavorites).toHaveBeenCalledWith(photos[0]);
    }));
  });
});
