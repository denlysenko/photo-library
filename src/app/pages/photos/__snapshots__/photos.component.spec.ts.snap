// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhotosComponent loadMore should call getPhotos() and render new photos 1`] = `
<app-photos
  cache={[Function Array]}
  loadMore={[Function Subject]}
  loading={[Function BehaviorSubject]}
  loading$={[Function Observable]}
  photos$={[Function Observable]}
  photosService={[Function Object]}
  route={[Function Object]}
>
  <app-infinite-scroll>
    <app-photo-grid>
      <div
        class="photo-grid"
      >
        
        <app-photo-tile
          data-testid="photo-tile"
        >
          <div
            class="photo-tile card sm"
          >
            <img
              alt="photo"
              class="photo"
              src="link"
            />
          </div>
        </app-photo-tile>
        <app-photo-tile
          data-testid="photo-tile"
        >
          <div
            class="photo-tile card sm"
          >
            <img
              alt="photo"
              class="photo"
              src="link2"
            />
          </div>
        </app-photo-tile>
      </div>
    </app-photo-grid>
    <app-inline-spinner
      class="hidden"
    >
      <div
        class="ellipsis"
      >
        <div />
        <div />
        <div />
        <div />
      </div>
    </app-inline-spinner>
    <div />
  </app-infinite-scroll>
</app-photos>
`;

exports[`PhotosComponent should create 1`] = `
<app-photos
  cache={[Function Array]}
  loadMore={[Function Subject]}
  loading={[Function BehaviorSubject]}
  loading$={[Function Observable]}
  photos$={[Function Observable]}
  photosService={[Function Object]}
  route={[Function Object]}
>
  <app-infinite-scroll>
    <app-photo-grid>
      <div
        class="photo-grid"
      >
        
        <app-photo-tile
          data-testid="photo-tile"
        >
          <div
            class="photo-tile card sm"
          >
            <img
              alt="photo"
              class="photo"
              src="link"
            />
          </div>
        </app-photo-tile>
      </div>
    </app-photo-grid>
    <app-inline-spinner
      class="hidden"
    >
      <div
        class="ellipsis"
      >
        <div />
        <div />
        <div />
        <div />
      </div>
    </app-inline-spinner>
    <div />
  </app-infinite-scroll>
</app-photos>
`;
